File: Encapsulation/CommissionEmployee/BasePlusCommissionEmployee.java
package Encapsulation.CommissionEmployee;

public class BasePlusCommissionEmployee extends CommissionEmployee {
    double  baseSalary;
    public BasePlusCommissionEmployee(String firstName, String lastName, String socialSecurityNumber, double grossSale, double commissionRate, double baseSalary){
        super(firstName, lastName, socialSecurityNumber, grossSale, commissionRate);
        setBaseSalary(baseSalary);
    }

    public double getBaseSalary() {
        return baseSalary;
    }


    public void setBaseSalary(double baseSalary) {
        if (baseSalary <= 0.0){
            throw new IllegalArgumentException("Base Salary should be greater than 0.0");
        }
        this.baseSalary = baseSalary;
    }

    @Override
    public double earnings(){
        return super.earnings() + baseSalary;
    }

    @Override
    public String toString(){
        return super.toString() + "\n Base Salary: " + baseSalary + "\n Earnings: " + earnings();
    }

    public static void main(String[] args) {
        BasePlusCommissionEmployee newEmployee = new BasePlusCommissionEmployee("Solomon", "Dzah", "22012447", 30, 0.4, 20);
        System.out.println(newEmployee.toString());

        // Updating the baseSalary and inherited fields (grossSales and commissionRate) 
        newEmployee.setGrossSales(40.5);
        newEmployee.setCommissionRate(0.51);

        // Displaying updated details
        System.out.println("Updated details: " + newEmployee.toString());

        // calculating and displaying total earnings
        double amount = newEmployee.earnings();
        System.out.println("Total Earnings: " + amount);

        // Testing the validation by attempting to set invalid values for baseSalary, grossSales,and commissionRate
        newEmployee.setBaseSalary(0);
        newEmployee.setGrossSales(-4.87);
        newEmployee.setCommissionRate(-0.876);
    }

}

File: python_lab_tasks/BasePlusCommissionEmployee.py
from CommissionEmployee import CommissionEmployee

class BasePlusCommissionEmployee(CommissionEmployee):
    def __init__(self, first_name, last_name, social_security_number, gross_sales, commission_rate, base_salary):
        super().__init__(first_name, last_name, social_security_number, gross_sales, commission_rate)
        if base_salary >= 0.0:
            self.__base_salary = base_salary
        else:
            raise ValueError("Base salary must be >= 0.0")

    def get_base_salary(self):
        return self.__base_salary

    def set_base_salary(self, value):
        if value >= 0.0:
            self.__base_salary = value
        else:
            raise ValueError("Base salary must be >= 0.0")

    def earnings(self):
        return self.__base_salary + super().earnings()

    def display_employee_details(self):
        return (super().display_employee_details() +
                f"\nBase Salary: {self.__base_salary}")


# Testing the derived class
if __name__ == "__main__":
    base_emp = BasePlusCommissionEmployee("Solomon", "Dzah", "592-194-480", 6000.0, 0.2, 1000.0)
    print(base_emp.display_employee_details())
    # Updating base salary and display updated earnings
    base_emp.set_base_salary(1200.0)
    print(f"Updated Base Salary: {base_emp.get_base_salary()}")
    print(f"Updated Earnings: {base_emp.earnings()}")
